### PR TITLE
Oracle Dialect

### PR DESCRIPTION
**Pull request by r0fls**
This is a WIP. I've been unable to get the fixture generated tests to parse. Also after installing it locally via `pip install -e .`, I get the following trying to parse a simple `COMMENT ON` statement:

```
$ pip install -e .
$ sqlfluff parse --dialect oracle foo.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        unparsable:                                           !! Expected: 'StatementSegment'
[L:  1, P:  1]      |            raw:                                              'COMMENT'
[L:  1, P:  8]      |            whitespace:                                       ' '
[L:  1, P:  9]      |            raw:                                              'ON'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            raw:                                              'TABLE'
[L:  1, P: 17]      |            whitespace:                                       ' '
[L:  1, P: 18]      |            raw:                                              'foo'
[L:  1, P: 21]      |            whitespace:                                       ' '
[L:  1, P: 22]      |            raw:                                              'IS'
[L:  1, P: 24]      |            whitespace:                                       ' '
[L:  1, P: 25]      |            raw:                                              "'a string'"
[L:  1, P: 35]      |    statement_terminator:                                     ';'
[L:  1, P: 36]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: "COMMENT ON TABLE
                       | foo IS 'a string'"
```

@jpy-git has been helping me in Slack, but I'm wondering if anyone can give me more pointers on what I'm doing wrong here since the implementation looks to similar enough to the Postgres `COMMENT ON` statement at this point to me, so I'm not sure what's missing. I will be adding other statements in future PRs after I get this one working.

FWIW, the test here is passing: https://github.com/sqlfluff/sqlfluff/pull/2293/files#diff-ab73b8d81c33cad7051b93c4dd1c4de6bfe6433f2dde88332e083e35fb86e1db

Related issue for this enhancement here: https://github.com/sqlfluff/sqlfluff/issues/2292

### CODE REVIEW

File: README.md
Comment (tunetheweb) [inline README.md:27]: Can you also add Oracle to docs/source/dialects.rst? Think it would be good to confirm scope of this dialect in there (SQL I presume and not PL/SQL? At least initially).

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_oracle.py:30]: This wouldn't be the way to do this. You would only "add" a segment like this if you want to add a recognisable "block".

For example:
```sql
-- This is a comment
```

This might be handled a as one "block", starting with `--` and then including any new line characters until then end of the line.

Here the `COMMENT` is just a keyword, as is covered above, so you don't need this code at all, so remove it.

```suggestion
```

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_oracle.py:24]: If removing `COMMENT` from `unreserved_keywords`, we should add it to `reserved_keywords` as well. also `ON` is already a reserved keyword in ANSI which you have inherited so don't need to add that again.

```suggestion
oracle_dialect.sets("unreserved_keywords").difference_update(["COMMENT"])
oracle_dialect.sets("reserved_keywords").update(["COMMENT"])
```

File: test/dialects/oracle_test.py
Comment (tunetheweb) [inline test/dialects/oracle_test.py:1]: Remove this file completely. We much prefer to add SQL examples to `test/fixtures/dialects/oracle`.

Create a `comment.sql` file in that directory and put the following in it:

```sql
COMMENT ON COLUMN employees.job_id 
   IS 'abbreviated job title';
```

Then run `tox -e generate-fixture-yml` which will generate a `comment.yml` file in same directory.

Add both to this branch and then both files are then automatically included in our test suite.

Managing SQL examples as test cases is MUCH easier than python code. We only revert to actual Python code for tests if there's a specific bit of Python code we want to test and can't test it with SQL statement but don;'t see that here so don't think we need this file.

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: README.md
Comment (r0fls) [reply README.md:27]: 🙏  I added a blurb. Let me know if that looks good or if I should add more clarity/context there.

**Comment by r0fls** [comment]
@tunetheweb thanks so much for the review! I had tried adding some tests via the fixtures method, but they weren't parsing correctly with the tox command. The new one is, so that's great!

However, I still see this when running through the manual installation (`pip install -e .`) and trying to test with a `foo.sql` file with the same content as the `.sql` file in the fixture 🤔 :

```
sqlfluff parse --dialect oracle foo.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        unparsable:                                           !! Expected: 'StatementSegment'
[L:  1, P:  1]      |            raw:                                              'COMMENT'
[L:  1, P:  8]      |            whitespace:                                       ' '
[L:  1, P:  9]      |            raw:                                              'ON'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            raw:                                              'COLUMN'
[L:  1, P: 18]      |            whitespace:                                       ' '
[L:  1, P: 19]      |            raw:                                              'employees'
[L:  1, P: 28]      |            raw:                                              '.'
[L:  1, P: 29]      |            raw:                                              'job_id'
[L:  1, P: 35]      |            whitespace:                                       ' '
[L:  1, P: 36]      |            newline:                                          '\n'
[L:  2, P:  1]      |            whitespace:                                       '   '
[L:  2, P:  4]      |            raw:                                              'IS'
[L:  2, P:  6]      |            whitespace:                                       ' '
[L:  2, P:  7]      |            raw:                                              "'abbreviated job title'"
[L:  2, P: 30]      |    statement_terminator:                                     ';'
[L:  2, P: 31]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'COMMENT ON COLUMN
                       | employees.job_id \n I...'
```

Any ideas what I'm doing wrong with the installation or usage here? I'd like to see it parse the file correctly too...

**Comment by tunetheweb** [comment]
Weird cause that worked for me parsing your fixture file directly:

```
% sqlfluff parse --dialect oracle test/fixtures/dialects/oracle/comment.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        comment_statement:
[L:  1, P:  1]      |            keyword:                                          'COMMENT'
[L:  1, P:  8]      |            whitespace:                                       ' '
[L:  1, P:  9]      |            keyword:                                          'ON'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            keyword:                                          'COLUMN'
[L:  1, P: 18]      |            whitespace:                                       ' '
[L:  1, P: 19]      |            column_reference:
[L:  1, P: 19]      |                identifier:                                   'employees'
[L:  1, P: 28]      |                dot:                                          '.'
[L:  1, P: 29]      |                identifier:                                   'job_id'
[L:  1, P: 35]      |            whitespace:                                       ' '
[L:  1, P: 36]      |            newline:                                          '\n'
[L:  2, P:  1]      |            whitespace:                                       '   '
[L:  2, P:  4]      |            keyword:                                          'IS'
[L:  2, P:  6]      |            whitespace:                                       ' '
[L:  2, P:  7]      |            literal:                                          "'abbreviated job title'"
[L:  2, P: 30]      |    statement_terminator:                                     ';'
[L:  2, P: 31]      |    newline:                                                  '\n'
```

Does that work for you?

File: docs/source/dialects.rst
Comment (tunetheweb) [inline docs/source/dialects.rst:91]:
```suggestion
.. _oracle_dialect_ref:
```

File: docs/source/dialects.rst
Comment (tunetheweb) [inline docs/source/dialects.rst:94]:
```suggestion
------
```

File: docs/source/dialects.rst
Comment (tunetheweb) [inline docs/source/dialects.rst:101]:
```suggestion
.. _postgres_dialect_ref:
```

**Review by tunetheweb — COMMENTED** [review]
Fix docs

File: test/fixtures/dialects/oracle/prompt-alter.sql
Comment (tunetheweb) [inline test/fixtures/dialects/oracle/prompt-alter.sql:8]: Maybe simplify these comments? Very long and confusing for test cases.

File: test/fixtures/dialects/oracle/prompt-alter.sql
Comment (tunetheweb) [inline test/fixtures/dialects/oracle/prompt-alter.sql:10]: This doesn't parse yet so remove for now until you get round to adding as it's failing the test suite.

```suggestion
```

File: test/fixtures/dialects/oracle/prompt-alter.yml
Comment (tunetheweb) [inline test/fixtures/dialects/oracle/prompt-alter.yml:124]: Here's where you see it failing. Want to check there are no `unparsable` section in the generated YML fails or the test suite won't complete.

```suggestion
```

Comment (r0fls) [reply test/fixtures/dialects/oracle/prompt-alter.yml]: Sweet, removed this file for now until I come back to prompt 💯

**Comment by r0fls** [comment]
>Weird cause that worked for me parsing your fixture file directly:

Still no 🤔 I'm thinking I must be doing the pip install incorrectly somehow, but it is accepting the `--dialect oracle` flag 🤔

If it works for you though, it must just be something I'm doing stupid...

**Comment by r0fls** [comment]
Removing WIP tag since we are getting closer to a useable first pass here.

**Comment by tunetheweb** [comment]
> > Weird cause that worked for me parsing your fixture file directly:
> 
> Still no... I'm thinking I must be doing the pip install incorrectly somehow, but it is accepting the `--dialect oracle` flag 🤔
> 
> If it works for you though, it must just be something I'm doing stupid.

Yes the fact it’s not complaining about oracle surely means it must be on your branch. That’s what’s really confusing me!!

Maybe start a refresh terminal and source again in case it’s some weird caching thing?

@jpy-git you’re good at this sort of thing. Any ideas?

> Removing WIP tag since we are getting closer to a useable first take here.

Yup. I’m a big fan of merging a new dialect early even if it seems barely different from ANSI. Easier to review smaller PRs and also allows others to contribute pieces then. So much better to do that IMHO than hold off until you’ve added all the statement your want - they can be follow on PRs.

Plus after your first merge you want need a maintainer to approve for the test suite to run in GitHub Actions!

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_oracle.py:60]: Looking at the docs it looks like `COMMENT ON VIEW ...` isn't a thing?
Both views and tables get use `COMMENT ON TABLE ...`.
Can you check this?
image

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_oracle.py:72]: Can we uncomment this so we have the complete grammar?
May need to add `INDEXTYPE` as a keyword.
I'm not sure if we have `IndexTypeReferenceSegment` in ANSI so may also need to create this (look at how `ColumnReferenceSegment` is defined for inspiration).

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_oracle.py:76]: I may be wrong here, but I feel we don't have a `MaterializedViewReferenceSegment`.

Referring to Postgres I can see we're using `TableReferenceSegment` like we do for normal views: https://github.com/sqlfluff/sqlfluff/blob/7257d289c08e8a2527117814c1f2d85e8c4f242f/src/sqlfluff/dialects/dialect_postgres.py#L1518

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_oracle.py:67]: Again don't think this is currently a thing.

We could either create this or could just use `ObjectReferenceSegment` if it's very niche.

File: test/fixtures/dialects/oracle/comment.sql
Comment (jpy-git) [inline test/fixtures/dialects/oracle/comment.sql:2]: Currently we only have a test case for `COMMENT ON COLUMN` here?

We should also have tests below for TABLE, INDEXTYPE, OPERATOR, and MATERIALIZED VIEW.

(n.b. as mentioned in other review comments, I think some of these other possible comment targets might not parse. Always best to have all cases covered in tests to prove that a PR works)

**Review by jpy-git — CHANGES_REQUESTED** [review]
Couple things to look into before merging

**Comment by jpy-git** [comment]
> Maybe start a refresh terminal and source again in case it’s some weird caching thing?
> 
> @jpy-git you’re good at this sort of thing. Any ideas?

hard to say tbh, maybe you had an outdated venv or some weird path setup 🤷 

If you're in a fresh virtual environment you should be fine. If you ever encounter issues with a venv the first step is always delete it and recreate:
tox:
```console
tox -e py39 -r --devenv .venv && source .venv/bin/activate
```
or
venv/pip:
```console
python3 -m venv .venv && source .venv/bin/activate && pip install -e .
```

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (r0fls) [reply src/sqlfluff/dialects/dialect_oracle.py]: Confirmed it is not a thing, thank you 🙏

File: test/fixtures/dialects/oracle/comment.sql
Comment (r0fls) [reply test/fixtures/dialects/oracle/comment.sql:2]: Yep, good point... easy coverage increase

File: src/sqlfluff/dialects/dialect_oracle.py
Comment (r0fls) [reply src/sqlfluff/dialects/dialect_oracle.py]: Went with `ObjectReferenceSegment` for now ✅

File: test/fixtures/dialects/oracle/comment.yml
Comment (jpy-git) [inline test/fixtures/dialects/oracle/comment.yml:19]: @r0fls Looks like you need to re-run `tox -e generate-fixture-yml` to add the include the extra unit tests in the fixture

Comment (r0fls) [reply test/fixtures/dialects/oracle/comment.yml:19]: Thank you, still getting `RuntimeError: Grammar refers to 'IndextypeKeywordSegment' which was not found in the oracle dialect` which is why I had it commented out 🤔 I grep'd around for `ObjectReferenceKeyword` and `TableReferenceKeyword` but couldn't find any examples of adding this to work from.

Comment (tunetheweb) [reply test/fixtures/dialects/oracle/comment.yml:19]: You need to ass `"INDEXTYPE"` to the keywords (in same way you added `"COMMENT"`,`"ON"`,`"UPDATE"`

Comment (r0fls) [reply test/fixtures/dialects/oracle/comment.yml:19]: Whoops, thank you! Added ✅

**Review by jpy-git — APPROVED** [review]
@r0fls Looks great, thanks for adding this! LGTM 🚀 ✨

**Comment by r0fls** [comment]
Thank you @jpy-git and @tunetheweb for the code review!

**Pull request merged**
